src/topics/async.md

```diff
@@ -0,0 +1,461 @@
+`setTimeout(..., 0)` is subtle.
+
+
+### Exceptions and Stack Traces
+Since an asynchronous function can call the callback some time after it was
+invoked, the execution environment of the callback can be quite different than
+what existed when the async function was called.
+
+Consider the following example:
+
+```js
+function err() {
+    var e = new Error('boom');
+    console.log(e.stack);
+}
+function run() {
+    err();
+    setTimeout(err, 100);
+}
+run();
+```
+
+This has the following output:
+
+```
+Error: boom
+    at err (/Users/drewfish/stacktrace.js:2:13)
+    at run (/Users/drewfish/stacktrace.js:6:5)
+    at Object.<anonymous> (/Users/drewfish/stacktrace.js:9:1)
+    at Module._compile (module.js:456:26)
+    at Object.Module._extensions..js (module.js:474:10)
+    at Module.load (module.js:356:32)
+    at Function.Module._load (module.js:312:12)
+    at Function.Module.runMain (module.js:497:10)
+    at startup (node.js:119:16)
+    at node.js:929:3
+Error: boom
+    at err [as _onTimeout] (/Users/drewfish/stacktrace.js:2:13)
+    at Timer.listOnTimeout [as ontimeout] (timers.js:112:15)
+```
+
+Note how the stacktrace from the first call to `err()` does have the `run()`
+function in it, yet the second call doesn't. This is because the second call
+to `err()` happens _after_ `run()` has completed so `run()` is no longer part
+of the execution environment.
+
+This is where the event loop shows up: the second call to `err()` happens
+during a timeout which is handled in the event loop.
+
+
+#### Throwing and try/catch
+One thing to be careful about in asynchronous programming is throwing errors
+and attempting to catch them.
+
+Consider this example:
+
+```js
+function run() {
+    setTimeout(function() {
+        throw new Error('boom');
+    }, 100);
+}
+try {
+    run();
+}
+catch (err) {
+    // we'll never get here!
+    console.log('TROUBLE!', err.stack);
+}
+```
+
+Since the `run()` function returns quickly, the execution of the program moves
+outside of the `try` block. (In fact it moves to the _end_ of the file, but
+the program doesn't exit because there is a timeout waiting to happen.) Once
+the timeout happens the code is run which throws the error. Since the program
+has moved outside of the `try` block the `catch` doesn't catch this thrown
+error.
+
+It is because of this that exceptions are not commonly thrown from
+asynchronous functions. Instead a common practice is to pass the error object
+to the callback. (See "nodeback" above.)
+
+In desperation, you can register a global exception handler by listening for
+the `uncaughtException` event on the process.
+
+```js
+process.on('uncaughtException', function(err) {
+  console.error('TROUBLE!', err.stack);
+});
+```
+
+Since this event handler is called well outside of your algorithm there is
+often very little you can do to recover from this exception.
+
+Some Promise implementations will catch a thrown error and resolve the promise
```

Comment:
Basically, what I said above - this needs some elaboration - and probably an example.